### ISSUE TITLE
`go_router` throws exception `initialLocation (locationHere) of StatefulShellBranch must be a valid location` when building a `StatefulShellRoute` tree with shared (mutual) routes in its branches, although different `NavigationState` keys are assigned to each branch.

### ISSUE THREAD

**Issue by MohaAmiry** [flutter/flutter#140753]
### Steps to reproduce

1- create 2 different `StatefulShellRoutes` and assigned 2 different `NavigationState` keys.
2- assign the same shared routes to them anywhere in the routing tree.

### Expected results

the expected result is: the shared routes must be accessible from both stacks.

### Actual results

exception because of a repeated route in a different shell: 
`
initialLocation (locationHere) of StatefulShellBranch must be a valid location
'package:go_router/src/configuration.dart':
Failed assertion: line 155 pos 17: '!matchList.isError'
`

### Code sample

<details open><summary>Code sample</summary>

the code below is a real case scenario where i needed to assign the same routes to different stacks (inserted a diagram image to represent the required flow): the customer have a 2 icons navigation bar (a stateful shell assigned for each icon), and he can view the shop routing tree immediately from the home route icon.
the company has a different 2 icons (shells) navigation bar, and can access the same shop routing tree but via different shell and different routes

```dart

final routerKey = GlobalKey<NavigatorState>(debugLabel: 'routerKey');

final _shellNavigatorCompanyKey1 = GlobalKey<NavigatorState>();
final _shellNavigatorCompanyKey2 = GlobalKey<NavigatorState>();

final _shellNavigatorCustomerKey1 = GlobalKey<NavigatorState>();
final _shellNavigatorCustomerKey2 = GlobalKey<NavigatorState>();

enum Routes {
  //shared routes' names
  servicesShopRoute("ShopService"),
  servicesShopSearchRoute("ShopServiceSearch"),
  servicesShopSelectedRoute("SelectedService"),

  //Company Routes
  companyCustomerOrdersRoute("/CustomerOrders"),
  companyMyServicesRoute("MyServices"),
  companyUniqueRoute("/SomeOtherRoute"),

  // customer routes
  customerHome('/CustomerHome'),
  customerProfile("/CustomerProfile");

  const Routes(this.path);

  final String path;
}

List<RouteBase> allRoutes = [
  StatefulShellRoute.indexedStack(
      builder: (context, state, navigationShell) {
        return NavigationBarViewSelection(
            // a widget that shows the dedicated navigation bar depending on the user type (customer,company)
            navigationShell: navigationShell,
            userType: UserType.consumer);
      },
      branches: [
        StatefulShellBranch(
            initialLocation: Routes.servicesShopRoute.path,
            navigatorKey: _shellNavigatorCustomerKey1,
            routes: [
              GoRoute(
                  path: Routes.customerHome.path,
                  builder: (context, state) => const Placeholder(),
                  routes: [
                    // ***************************************************************
                    // shared Routes
                    GoRoute(
                        path: Routes.servicesShopRoute.path,
                        builder: (context, state) => const Placeholder(),
                        routes: [
                          GoRoute(
                              path: Routes.servicesShopSearchRoute.path,
                              builder: (context, state) => const Placeholder()),
                          GoRoute(
                              path: Routes.servicesShopSelectedRoute.path,
                              builder: (context, state) => const Placeholder())
                        ]),
                    //****************************************************************
                  ]),
            ]),
        StatefulShellBranch(
            initialLocation: Routes.customerProfile.path,
            navigatorKey: _shellNavigatorCustomerKey2,
            routes: [
              GoRoute(
                  path: Routes.customerProfile.path,
                  builder: (context, state) => const Placeholder()),
            ])
      ]),
  StatefulShellRoute.indexedStack(
      builder: (context, state, navigationShell) {
        return NavigationBarViewSelection(
            navigationShell: navigationShell, userType: UserType.guest);
      },
      branches: [
        StatefulShellBranch(
            initialLocation: Routes.companyCustomerOrdersRoute.path,
            navigatorKey: _shellNavigatorCompanyKey1,
            routes: [
              GoRoute(
                  path: Routes.companyCustomerOrdersRoute.path,
                  builder: (context, state) => const Placeholder(),
                  routes: [
                    GoRoute(
                        path: Routes.companyMyServicesRoute.path,
                        builder: (context, state) => const Placeholder(),
                        routes: [
                          // ***************************************************************
                          // shared Routes
                          GoRoute(
                              path: Routes.servicesShopRoute.path,
                              builder: (context, state) => const Placeholder(),
                              routes: [
                                GoRoute(
                                    path: Routes.servicesShopSearchRoute.path,
                                    builder: (context, state) =>
                                        const Placeholder()),
                                GoRoute(
                                    path: Routes.servicesShopSelectedRoute.path,
                                    builder: (context, state) =>
                                        const Placeholder())
                              ]),
                          //****************************************************************
                        ])
                  ]),
            ]),
        StatefulShellBranch(
            initialLocation: Routes.companyUniqueRoute.path,
            navigatorKey: _shellNavigatorCompanyKey2,
            routes: [
              GoRoute(
                  path: Routes.companyUniqueRoute.path,
                  builder: (context, state) => const Placeholder())
            ]),
      ])
];

var router = GoRouter(
    navigatorKey: routerKey,
    initialLocation: Routes.customerHome.path,
    debugLogDiagnostics: true,
    routes: allRoutes);

void main() async {
  WidgetsFlutterBinding.ensureInitialized();

  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp.router(routerConfig: router);
  }
}


```

</details>

### Screenshots or Video

<details open>
<summary>Screenshots / Video demonstration</summary>
a real case scenario is like this: the customer have a 2 icons navigation bar (a stateful shell assigned for each icon), and he can view the shop routing tree immediately from the home route icon.
the company has a different 2 icons (shells) navigation bar, and can access the same shop routing tree but via different shell and different routes.

![Asset 1@3x-100](https://github.com/flutter/flutter/assets/65380552/e12de441-bd45-4539-88e6-8a1ee6cfcdcb)

</details>

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
[√] Flutter (Channel stable, 3.16.3, on Microsoft Windows [Version 10.0.22631.2861], locale en-US)
    • Flutter version 3.16.3 on channel stable at C:\flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision b0366e0a3f (3 weeks ago), 2023-12-05 19:46:39 -0800
    • Engine revision 54a7145303
    • Dart version 3.2.3
    • DevTools version 2.28.4

[√] Windows Version (Installed version of Windows is version 10 or higher)

[√] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at C:\Users\mdbm\AppData\Local\Android\sdk
    • Platform android-34, build-tools 34.0.0
    • Java binary at: C:\Program Files\Android\Android Studio\jbr\bin\java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-b2043.56-10027231)
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • CHROME_EXECUTABLE = C:\Program Files (x86)\Microsoft\Edge\Application\msedge.exe

[√] Visual Studio - develop Windows apps (Visual Studio Enterprise 2022 17.7.4)
    • Visual Studio at C:\Program Files\Microsoft Visual Studio\2022\Enterprise
    • Visual Studio Enterprise 2022 version 17.7.34031.279
    • Windows 10 SDK version 10.0.22621.0

[√] Android Studio (version 2022.3)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-b2043.56-10027231)

[√] IntelliJ IDEA Ultimate Edition (version 2023.2)
    • IntelliJ at C:\Program Files\JetBrains\IntelliJ IDEA 2023.2
    • Flutter plugin version 75.1.4
    • Dart plugin version 232.8660.129

[√] Connected device (4 available)
    • sdk gphone64 x86 64 (mobile) • emulator-5554 • android-x64    • Android 14 (API 34) (emulator)
    • Windows (desktop)            • windows       • windows-x64    • Microsoft Windows [Version 10.0.22631.2861]
    • Chrome (web)                 • chrome        • web-javascript • unknown
    • Edge (web)                   • edge          • web-javascript • Microsoft Edge 120.0.2210.91

[√] Network resources
    • All expected network resources are available.

• No issues found!
```

</details>

**Comment by dam-ease** [comment]
Thanks for filing this @MohaAmiry.
I can reproduce this on the latest `master` and `stable` channels as well.

<details><summary>Code Sample Used</summary>
<p>

```dart
import 'package:flutter/material.dart';
import 'package:go_router/go_router.dart';

final routerKey = GlobalKey<NavigatorState>(debugLabel: 'routerKey');

final _shellNavigatorCompanyKey1 = GlobalKey<NavigatorState>();
final _shellNavigatorCompanyKey2 = GlobalKey<NavigatorState>();

final _shellNavigatorCustomerKey1 = GlobalKey<NavigatorState>();
final _shellNavigatorCustomerKey2 = GlobalKey<NavigatorState>();

enum Routes {
  //shared routes' names
  servicesShopRoute("ShopService"),
  servicesShopSearchRoute("ShopServiceSearch"),
  servicesShopSelectedRoute("SelectedService"),

  //Company Routes
  companyCustomerOrdersRoute("/CustomerOrders"),
  companyMyServicesRoute("MyServices"),
  companyUniqueRoute("/SomeOtherRoute"),

  // customer routes
  customerHome('/CustomerHome'),
  customerProfile("/CustomerProfile");

  const Routes(this.path);

  final String path;
}

List<RouteBase> allRoutes = [
  StatefulShellRoute.indexedStack(
      builder: (context, state, navigationShell) {
        return Container();

        // return NavigationBarViewSelection(
        //     // a widget that shows the dedicated navigation bar depending on the user type (customer,company)
        //     navigationShell: navigationShell,
        //     userType: UserType.consumer);
      },
      branches: [
        StatefulShellBranch(
            initialLocation: Routes.servicesShopRoute.path,
            navigatorKey: _shellNavigatorCustomerKey1,
            routes: [
              GoRoute(
                  path: Routes.customerHome.path,
                  builder: (context, state) => const Placeholder(),
                  routes: [
                    // ***************************************************************
                    // shared Routes
                    GoRoute(
                        path: Routes.servicesShopRoute.path,
                        builder: (context, state) => const Placeholder(),
                        routes: [
                          GoRoute(
                              path: Routes.servicesShopSearchRoute.path,
                              builder: (context, state) => const Placeholder()),
                          GoRoute(
                              path: Routes.servicesShopSelectedRoute.path,
                              builder: (context, state) => const Placeholder())
                        ]),
                    //****************************************************************
                  ]),
            ]),
        StatefulShellBranch(
            initialLocation: Routes.customerProfile.path,
            navigatorKey: _shellNavigatorCustomerKey2,
            routes: [
              GoRoute(
                  path: Routes.customerProfile.path,
                  builder: (context, state) => const Placeholder()),
            ])
      ]),
  StatefulShellRoute.indexedStack(
      builder: (context, state, navigationShell) {
        return Container();
        // return NavigationBarViewSelection(
        //     navigationShell: navigationShell, userType: UserType.guest);
      },
      branches: [
        StatefulShellBranch(
            initialLocation: Routes.companyCustomerOrdersRoute.path,
            navigatorKey: _shellNavigatorCompanyKey1,
            routes: [
              GoRoute(
                  path: Routes.companyCustomerOrdersRoute.path,
                  builder: (context, state) => const Placeholder(),
                  routes: [
                    GoRoute(
                        path: Routes.companyMyServicesRoute.path,
                        builder: (context, state) => const Placeholder(),
                        routes: [
                          // ***************************************************************
                          // shared Routes
                          GoRoute(
                              path: Routes.servicesShopRoute.path,
                              builder: (context, state) => const Placeholder(),
                              routes: [
                                GoRoute(
                                    path: Routes.servicesShopSearchRoute.path,
                                    builder: (context, state) =>
                                        const Placeholder()),
                                GoRoute(
                                    path: Routes.servicesShopSelectedRoute.path,
                                    builder: (context, state) =>
                                        const Placeholder())
                              ]),
                          //****************************************************************
                        ])
                  ]),
            ]),
        StatefulShellBranch(
            initialLocation: Routes.companyUniqueRoute.path,
            navigatorKey: _shellNavigatorCompanyKey2,
            routes: [
              GoRoute(
                  path: Routes.companyUniqueRoute.path,
                  builder: (context, state) => const Placeholder())
            ]),
      ])
];

var router = GoRouter(
    navigatorKey: routerKey,
    initialLocation: Routes.customerHome.path,
    debugLogDiagnostics: true,
    routes: allRoutes);

void main() async {
  WidgetsFlutterBinding.ensureInitialized();

  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp.router(routerConfig: router);
  }
}

``` 

</p>
</details> 

<details><summary>Logs</summary>
<p>

```
#53     RootElement._rebuild (package:flutter/src/widgets/binding.dart:1354:16)
binding.dart:1354
#54     RootElement.mount (package:flutter/src/widgets/binding.dart:1323:5)
binding.dart:1323
#55     RootWidget.attach.<anonymous closure> (package:flutter/src/widgets/binding.dart:1276:18)
binding.dart:1276
#56     BuildOwner.buildScope (package:flutter/src/widgets/framework.dart:2844:19)
framework.dart:2844
#57     RootWidget.attach (package:flutter/src/widgets/binding.dart:1275:13)
binding.dart:1275
#58     WidgetsBinding.attachToBuildOwner (package:flutter/src/widgets/binding.dart:1088:27)
binding.dart:1088
#59     WidgetsBinding.attachRootWidget (package:flutter/src/widgets/binding.dart:1070:5)
binding.dart:1070
#60     WidgetsBinding.scheduleAttachRootWidget.<anonymous closure> (package:flutter/src/widgets/binding.dart:1056:7)
binding.dart:1056
#64     _RawReceivePort._handleMessage (dart:isolate-patch/isolate_patch.dart:184:12)
isolate_patch.dart:184
(elided 5 frames from class _AssertionError, class _Timer, and dart:async-patch)

════════════════════════════════════════════════════════════════════════════════
``` 

</p>
</details> 

<details><summary>stable, master flutter doctor -v</summary>
<p>

```
[!] Flutter (Channel stable, 3.16.5, on macOS 14.2.1 23C71 darwin-arm64, locale
    en-NG)
    • Flutter version 3.16.5 on channel stable at
      /Users/damilolaalimi/sdks/flutter
    ! Warning: `dart` on your path resolves to
      /opt/homebrew/Cellar/dart/3.1.5/libexec/bin/dart, which is not inside your
      current Flutter SDK checkout at /Users/damilolaalimi/sdks/flutter.
      Consider adding /Users/damilolaalimi/sdks/flutter/bin to the front of your
      path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 78666c8dc5 (13 days ago), 2023-12-19 16:14:14 -0800
    • Engine revision 3f3e560236
    • Dart version 3.2.3
    • DevTools version 2.28.4
    • If those were intentional, you can disregard the above warnings; however
      it is recommended to use "git" directly to perform update checks and
      upgrades.

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /Users/damilolaalimi/Library/Android/sdk
    • Platform android-34, build-tools 34.0.0
    • ANDROID_HOME = /Users/damilolaalimi/Library/Android/sdk
    • Java binary at: /Applications/Android
      Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      17.0.6+0-17.0.6b802.4-9586694)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 15C65
    • CocoaPods version 1.14.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2022.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build
      17.0.6+0-17.0.6b802.4-9586694)

[!] Android Studio (version unknown)
    • Android Studio at /Users/damilolaalimi/Downloads/Android Studio
      Preview.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    ✗ Unable to determine Android Studio version.
    • Java version OpenJDK Runtime Environment (build
      17.0.7+0-17.0.7b1000.6-10550314)

[✓] VS Code (version 1.85.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.78.0

[✓] VS Code (version 1.83.1)
    • VS Code at /Users/damilolaalimi/Downloads/Visual Studio Code.app/Contents
    • Flutter extension version 3.78.0

[✓] Connected device (3 available)
    • Damilola’s iPhone (mobile) • 00008110-001964480AE1801E • ios            •
      iOS 17.1.1 21B91
    • macOS (desktop)            • macos                     • darwin-arm64   •
      macOS 14.2.1 23C71 darwin-arm64
    • Chrome (web)               • chrome                    • web-javascript •
      Google Chrome 120.0.6099.129

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 2 categories.
``` 
```
[!] Flutter (Channel master, 3.18.0-18.0.pre.47, on macOS 14.2.1 23C71 darwin-arm64, locale en-NG)
    • Flutter version 3.18.0-18.0.pre.47 on channel master at /Users/damilolaalimi/fvm/versions/master
    ! Warning: `dart` on your path resolves to /opt/homebrew/Cellar/dart/3.1.5/libexec/bin/dart, which is not inside your current Flutter SDK checkout at /Users/damilolaalimi/fvm/versions/master. Consider adding /Users/damilolaalimi/fvm/versions/master/bin to the front of your path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 3df4e735d5 (4 days ago), 2023-12-29 14:07:39 -0500
    • Engine revision 9e03a57cde
    • Dart version 3.3.0 (build 3.3.0-273.0.dev)
    • DevTools version 2.31.0-dev.0
    • If those were intentional, you can disregard the above warnings; however it is recommended to use "git" directly to perform update checks and upgrades.

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /Users/damilolaalimi/Library/Android/sdk
    • Platform android-34, build-tools 34.0.0
    • ANDROID_HOME = /Users/damilolaalimi/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 15C65
    • CocoaPods version 1.14.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2022.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)

[!] Android Studio (version unknown)
    • Android Studio at /Users/damilolaalimi/Downloads/Android Studio Preview.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    ✗ Unable to determine Android Studio version.
    • Java version OpenJDK Runtime Environment (build 17.0.7+0-17.0.7b1000.6-10550314)

[✓] VS Code (version 1.85.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.78.0

[✓] VS Code (version 1.83.1)
    • VS Code at /Users/damilolaalimi/Downloads/Visual Studio Code.app/Contents
    • Flutter extension version 3.78.0

[✓] Connected device (3 available)
    • Damilola’s iPhone (mobile) • 00008110-001964480AE1801E • ios            • iOS 17.1.1 21B91
    • macOS (desktop)            • macos                     • darwin-arm64   • macOS 14.2.1 23C71 darwin-arm64
    • Chrome (web)               • chrome                    • web-javascript • Google Chrome 120.0.6099.129

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 2 categories.
``` 

</p>
</details>

**Comment by hangyujin** [comment]
Hi @MohaAmiry , I'm using your code example to reproduce the exception, it doesn't seem to be related to shared routes, even if you remove the second StatefulShellRoute the exception is still there. 

I think this exception is just related to `initialLocation`.

Changing the line`initialLocation: Routes.servicesShopRoute.path` to `initialLocation: Routes.customerHome.path` will fix the issue.
